coordinator/points_writer.go

```diff
@@ -195,57 +196,90 @@ func (w *PointsWriter) Statistics(tags map[string]string) []models.Statistic {
 // maps to a shard group or shard that does not currently exist, it will be
 // created before returning the mapping.
 func (w *PointsWriter) MapShards(wp *WritePointsRequest) (*ShardMapping, error) {
-
-	// holds the start time ranges for required shard groups
-	timeRanges := map[time.Time]*meta.ShardGroupInfo{}
-
 	rp, err := w.MetaClient.RetentionPolicy(wp.Database, wp.RetentionPolicy)
 	if err != nil {
 		return nil, err
 	} else if rp == nil {
 		return nil, influxdb.ErrRetentionPolicyNotFound(wp.RetentionPolicy)
 	}
 
-	// Find the minimum time for a point if the retention policy has a shard
-	// group duration. We will automatically drop any points before this time.
-	// There is a chance of a time on the edge of the shard group duration to
-	// sneak through even after it has been removed, but the circumstances are
-	// rare enough and don't matter enough that we don't account for this
-	// edge case.
-	min := time.Unix(0, models.MinNanoTime)
-	if rp.Duration > 0 {
-		min = time.Now().Add(-rp.Duration)
-	}
-
+	// Holds all the shard groups and shards that are required for writes.
+	var list sgList
```

Comment:
This should probably start with some size, maybe 4 or 8? Generally we'll only see one, or two when we're near a shard boundary, but a few extra shouldn't hurt too much. It might even be beneficial in the long-run to pool these slices to reduce the churn we'll see.